### PR TITLE
Fix edge case where service interface type could not be determined and improve tests

### PR DESCRIPTION
**Pull request by FejZa**
# Reality Collective - Reality Toolkit Pull Request

## Overview

Adds a bunch of tests to ensure service retrieval using a base type works as expected. Also fixes another edge case when looking up the interface type for a service to be registered.

## Changes

What I said above.

## Breaking Changes

None.

## Related Submodule Changes

None

## Testing status

Includes unit tests.

### Manual testing status

Tested and verified in custom project with increased complexity and many custom services and providers.

### CODE REVIEW

File: Runtime/Services/ServiceManager.cs [hunk 1]
```diff
@@ -979,16 +979,20 @@ public bool TryGetService(Type interfaceType, string serviceName, out IService s
 
             if (!CanGetService(interfaceType, serviceName)) { return false; }
 
-            if (activeServices.TryGetValue(interfaceType, out var service))
+            foreach (var activeService in activeServices)
```
Comment (SimonDarksideJ) [inline Runtime/Services/ServiceManager.cs:982]: TryGetValue performs this same operation but in a safe context and also handles if "activeServices" is null which this does not.

Any reason to change this?
As far as I can see, this is just reordering the AssignableFrom and CheckService, so maybe just to that instead of switching to an unsafe foreach.

File: Tests/TestData/Interfaces/ITestService.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using RealityCollective.ServiceFramework.Interfaces;
+
+namespace RealityCollective.ServiceFramework.Tests.Interfaces
+{
+    /// <summary>
+    /// Common base interface for all services that are used for testing only.
+    /// </summary>
+    public interface ITestService : IService { }
```
Comment (SimonDarksideJ) [inline Tests/TestData/Interfaces/ITestService.cs:11]: If testing for the scenario of nested interfaces, I would have expected a few more test classes with that behaviour
E.G. 
TestService : MyServiceBase -> MyServiceBase : MyInterface -> MyInterface : IService

**Review by SimonDarksideJ — CHANGES_REQUESTED** [review]

File: Runtime/Services/ServiceManager.cs (hunk 1 above)
Comment (FejZa) [reply Runtime/Services/ServiceManager.cs:982]: TryGetValue is nowhere the same as my change. Please take a look again what it does. TryGetValue will look for a specific value in the dictionary, which will obviously always return false for anything that is not in the dictionary. Whereas the new solution will ensure that interface inheritance is respected and return true, if the value in the dictionary is assignable for the requested interface type.

Without this change those tests and scenarios will fail, with the change things work as expected.

Comment (FejZa) [reply Runtime/Services/ServiceManager.cs:982]: I don't see any null check there for activeServices. TryGetValue does not check, whether activeServices is null and would raise an NRE if it was. activeServices must never be null, that'd be a major design flaw.

Comment (FejZa) [reply Runtime/Services/ServiceManager.cs:982]: The least is it supposed to be empty.

Comment (FejZa) [reply Runtime/Services/ServiceManager.cs:982]: I don't feel like you took the time to understand what the change does tbh.

**Review by SimonDarksideJ — APPROVED** [review]
